llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp

```diff
@@ -1333,6 +1308,56 @@ static void addPreloadKernArgHint(Function &F, TargetMachine &TM) {
   }
 }
 
+/// The final check and update of the attribute 'amdgpu-waves-per-eu' based on
+/// the determined 'amdgpu-flat-work-group-size' attribute. We can't do this
+/// during attributor run because the two attributes grow in opposite direction,
+/// we should not use any intermediate value to calculate waves per eu until we
+/// have a determined flat workgroup size.
+static void updateWavesPerEU(Module &M, TargetMachine &TM) {
```

Comment:
It just occurred to me that AMDGPUAttributor is only a Module pass. Should we add or replace it with a CGSCC pass?

Reply:
We should probably add a CGSCC pass.